kernel/src/process.rs

```diff
@@ -1143,43 +1290,190 @@ impl<C: Chip> ProcessType for Process<'_, C> {
+            // memory (verified using `in_app_owned_memory`) and
+            // respect alignment and other constraints of the Rust
+            // references created by ReadWriteAppSlice.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, size, self.appid()) }
+        } else {
+            return GenericSyscallReturnValue::AllowReadWriteFailure(
+                ErrorCode::INVAL,
+                buf_start_addr,
+                size,
+            );
+        };
+
+        let allow_result = driver_invoc_closure(new_slice);
+
+        // TODO: Check whether the capsule either made the allow fail
+        // and returned the new slice, or it made the allow succeed
+        // and returned the expected slice from the allow table
+
+        match allow_result {
+            Ok(old_slice) => {
+                let (ptr, len) = old_slice.consume();
+                GenericSyscallReturnValue::AllowReadWriteSuccess(ptr, len)
+            }
+            Err((new_slice, err)) => {
+                let (ptr, len) = new_slice.consume();
+                GenericSyscallReturnValue::AllowReadWriteFailure(err, ptr, len)
+            }
+        }
+    }
+
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
+    fn allow_readonly(
         &self,
         buf_start_addr: *const u8,
         size: usize,
-    ) -> Result<Option<AppSlice<Shared, u8>>, ReturnCode> {
+        driver_invoc_closure: &dyn Fn(
+            ReadOnlyAppSlice,
+        )
+            -> Result<ReadOnlyAppSlice, (ReadOnlyAppSlice, ErrorCode)>,
+    ) -> GenericSyscallReturnValue {
         if !self.is_active() {
-            // Do not modify an inactive process.
-            return Err(ReturnCode::FAIL);
+            // Do not operate on an inactive process
+            return GenericSyscallReturnValue::AllowReadOnlyFailure(
+                ErrorCode::FAIL,
+                buf_start_addr,
+                size,
+            );
         }
 
-        match NonNull::new(buf_start_addr as *mut u8) {
-            None => {
-                // A null buffer means pass in `None` to the capsule
-                Ok(None)
+        // A process is allowed to pass any pointer if the slice
+        // length is 0, as to revoke kernel access to a memory region
+        // without granting access to another one
+        let new_slice = if size == 0 {
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as a slice
+            // of length 0 will never allow dereferencing any memory
+            // in a safe manner.
```

Comment:
Yeah, that's what I thought at least. Turns out, memory safety isn't as easy as that. Going to fix this comment along with the actual fix for the bug introduced by this assumption in #2464.

Reply:
marking resolved because this is covered in #2464